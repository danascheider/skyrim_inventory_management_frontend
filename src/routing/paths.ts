export type PathString = `/${string}`

interface Paths {
  [i: string]: PathString
}

const paths: Paths = {
  home: '/',
  login: '/login'
}

export default paths
