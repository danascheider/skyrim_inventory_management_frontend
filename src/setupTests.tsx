import { type ReactElement } from 'react'
import { BrowserRouter } from 'react-router-dom'
import { render } from '@testing-library/react'

export const renderWithRouter = (ui: ReactElement) => render(<BrowserRouter>{ui}</BrowserRouter>)