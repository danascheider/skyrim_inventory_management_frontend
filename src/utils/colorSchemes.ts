export interface ColorScheme {
  schemeColorDarkest: string
  schemeColorDark: string
  schemeColorLight: string
  schemeColorLightest: string
  hoverColorDark: string
  hoverColorLight: string
  hoverColorLightest: string
  textColorPrimary: string
  textColorSecondary: string
  textColorTertiary: string
  borderColor: string
}

export const YELLOW: ColorScheme = {
  schemeColorDarkest: '#ffbf00',
  schemeColorDark: '#ffcb32',
  schemeColorLight: '#ffd866',
  schemeColorLightest: '#fff2cc',
  hoverColorDark: '#e5ab00',
  hoverColorLight: '#ffc519',
  hoverColorLightest: '#ffebb2',
  textColorPrimary: '#4c3900',
  textColorSecondary: '#4c3900',
  textColorTertiary: '#7f5700',
  borderColor: '#cc9800',
}

export const PINK: ColorScheme = {
  schemeColorDarkest: '#e83f6f',
  schemeColorDark: '#ec658b',
  schemeColorLight: '#f18ba8',
  schemeColorLightest: '#fad8e2',
  hoverColorDark: '#d03863',
  hoverColorLight: '#ea527d',
  hoverColorLightest: '#f5b2c5',
  textColorPrimary: '#fff',
  textColorSecondary: '#fff',
  textColorTertiary: '#451221',
  borderColor: '#b93258',
}

export const BLUE: ColorScheme = {
  schemeColorDarkest: '#2274a5',
  schemeColorDark: '#4e8fb7',
  schemeColorLight: '#7aabc9',
  schemeColorLightest: '#d2e3ed',
  hoverColorDark: '#1e6894',
  hoverColorLight: '#3881ae',
  hoverColorLightest: '#bcd5e4',
  textColorPrimary: '#fff',
  textColorSecondary: '#fff',
  textColorTertiary: '#0d2e42',
  borderColor: '#1b5c84',
}

export const GREEN: ColorScheme = {
  schemeColorDarkest: '#00a323',
  schemeColorDark: '#32b54e',
  schemeColorLight: '#66c77b',
  schemeColorLightest: '#ccecd3',
  hoverColorDark: '#00921f',
  hoverColorLight: '#19ac38',
  hoverColorLightest: '#b2e3bd',
  textColorPrimary: '#fff',
  textColorSecondary: '#fff',
  textColorTertiary: '#00410e',
  borderColor: '#00821c',
}

export const AQUA: ColorScheme = {
  schemeColorDarkest: '#20e2e9',
  schemeColorDark: '#62eaef',
  schemeColorLight: '#8ff0f4',
  schemeColorLightest: '#d2f9fa',
  hoverColorDark: '#11cbd1',
  hoverColorLight: '#4ce7ed',
  hoverColorLightest: '#bcf6f8',
  textColorPrimary: '#094345',
  textColorSecondary: '#094345',
  textColorTertiary: '#094345',
  borderColor: '#19b4ba',
}

const colorSchemes: ColorScheme[] = [YELLOW, PINK, BLUE, GREEN, AQUA]

export default colorSchemes
