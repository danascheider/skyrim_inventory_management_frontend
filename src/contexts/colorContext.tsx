import { createContext, ReactElement } from 'react'
import { YELLOW, type ColorScheme } from '../utils/colorSchemes'

const ColorContext = createContext<ColorScheme>(YELLOW)

interface ProviderProps {
  colorScheme: ColorScheme
  children: ReactElement
}

const ColorProvider = ({ colorScheme, children }: ProviderProps) => (
  <ColorContext.Provider value={colorScheme}>{children}</ColorContext.Provider>
)

export { ColorProvider, ColorContext }
