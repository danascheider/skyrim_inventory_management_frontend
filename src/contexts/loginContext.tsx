import { createContext, useState, useEffect, useCallback } from 'react'
import { type User } from 'firebase/auth'
import { useAuthState } from 'react-firebase-hooks/auth'
import { auth, signOutWithGoogle } from '../firebase'
import { ProviderProps } from '../types/contexts'

export interface LoginContextType {
  authLoading: boolean
  token: string | null
  withTokenRefresh: (fn: (idToken: string) => void) => void
  user?: User | null
}

export const LoginContext = createContext<LoginContextType>({
  user: null,
  token: null,
  authLoading: true,
  withTokenRefresh: () => {},
})

export const LoginProvider = ({ children }: ProviderProps) => {
  const [user, authLoading, authError] = useAuthState(auth)
  const [token, setToken] = useState<string | null>(null)

  const refreshToken = useCallback(() => {
    if (user) user.getIdToken(true).then((idToken) => setToken(idToken))
  }, [user])

  const withTokenRefresh = (fn: Function) => {
    if (user) {
      user.getIdToken(true).then((idToken) => {
        setToken(idToken)
        fn(idToken)
      })
    }
  }

  const value = {
    user,
    token,
    withTokenRefresh,
    authLoading,
  }

  useEffect(() => {
    if (authError) signOutWithGoogle()

    refreshToken()
  }, [user, authError])

  return <LoginContext.Provider value={value}>{children}</LoginContext.Provider>
}
