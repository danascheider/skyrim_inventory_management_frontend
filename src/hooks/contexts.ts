import { useContext } from 'react'
import { ColorContext } from '../contexts/colorContext'
import { LoginContext } from '../contexts/loginContext'

const useCustomContext = <T>(cxt: React.Context<T>, msg: string) => {
  const context = useContext(cxt)

  if (!context) throw new Error(msg)

  return context
}

export const useColorScheme = () =>
  useCustomContext(
    ColorContext,
    'useColorScheme must be used within a ColorProvider'
  )

export const useGoogleLogin = () =>
  useCustomContext(
    LoginContext,
    'useGoogleLogin must be used within a LoginProvider'
  )