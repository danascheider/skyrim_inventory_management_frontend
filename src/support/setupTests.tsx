import { type ReactElement } from 'react'
import { BrowserRouter } from 'react-router-dom'
import { render } from '@testing-library/react'
import { LoginContext } from '../contexts/loginContext'
import { User } from 'firebase/auth'
import testProfileImg from './testProfileImg.png'

export const requireLogin = () => { /* noop */ }

export const testUser = {
  uid: 'edna',
  displayName: 'Edna St. Vincent Millay',
  email: 'edna@gmail.com',
  photoURL: testProfileImg,
  emailVerified: true,
  getIdToken: () => new Promise<string>((resolve, _reject) => resolve('xxxxxxx'))
} as User

export const renderWithRouter = (ui: ReactElement) =>
  render(<BrowserRouter>{ui}</BrowserRouter>)

export const renderAuthenticated = (
  ui: ReactElement,
  authLoading = false,
  authError = undefined
) =>
  renderWithRouter(
    <LoginContext.Provider value={{ user: testUser, token: 'xxxxxxx', requireLogin, authLoading, authError }}>
      {ui}
    </LoginContext.Provider>
  )

export const renderUnauthenticated = (ui: ReactElement) =>
  renderWithRouter(
    <LoginContext.Provider value={{ user: null, token: null, authLoading: false, requireLogin }}>
      {ui}
    </LoginContext.Provider>
  )

export const renderAuthLoading = (ui: ReactElement) =>
  renderWithRouter(
    <LoginContext.Provider value={{ user: null, token: null, authLoading: true, requireLogin }}>
      {ui}
    </LoginContext.Provider>
  )
