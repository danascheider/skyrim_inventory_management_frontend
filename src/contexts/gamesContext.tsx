import { createContext, useCallback, useEffect, useState } from 'react'
import { signOutWithGoogle } from '../firebase'
import { type Game } from '../types/games'
import { type ProviderProps } from '../types/contexts'
import { useGoogleLogin, usePageContext } from '../hooks/contexts'
import { ApiError } from '../types/errors'
import { LOADING, DONE, ERROR, type LoadingState } from '../utils/loadingStates'
import { getGames, deleteGame } from '../utils/simApi'

// TODO: Add default values for context provider as follows:
//       const createGame = (game: Game) => {
//                                            postGames(game: Game, token: '')
//                                              .then((status: number, json: string) => /* noop */)
//                                          }
//       This will be useful in Storybook.

const NOT_FOUND_MESSAGE =
  "Oops! We couldn't find the game you're looking for. Please refresh and try again."
const UNEXPECTED_ERROR_MESSAGE =
  "Oops! Something unexpected went wrong. We're sorry! Please try again later."

export interface GamesContextType {
  games: Game[]
  gamesLoadingState: LoadingState
  // createGame: (game: Game) => void
  // updateGame: (gameId: number, attrs: Game) => void
  destroyGame: (gameId: number) => void
}

export const GamesContext = createContext<GamesContextType>({
  games: [],
  gamesLoadingState: LOADING,
  destroyGame: () => {},
})

export const GamesProvider = ({ children }: ProviderProps) => {
  const { user, token, authLoading, requireLogin } = useGoogleLogin()
  const [gamesLoadingState, setGamesLoadingState] = useState(LOADING)
  const [games, setGames] = useState<Game[]>([])
  const { setFlashProps } = usePageContext()

  /*
   *
   * Retrieve all the current user's games from the API
   *
   */

  const fetchGames = useCallback(() => {
    if (user && token) {
      getGames(token)
        .then(({ json }) => {
          if (Array.isArray(json)) {
            setGames(json)
            setGamesLoadingState(DONE)
          }
        })
        .catch((e: ApiError) => {
          if (import.meta.env.DEV)
            console.error(`Error ${e.name}: ${e.message}`)

          if (e.code === 401) {
            signOutWithGoogle()
          } else {
            setGames([])
            setGamesLoadingState(ERROR)
            setFlashProps({
              type: 'error',
              message: e.message,
              hidden: false,
            })
          }
        })
    }
  }, [user, token])

  /*
   *
   * Destroy the requested game and update the `games` array
   *
   */

  const destroyGame = useCallback(
    (gameId: number) => {
      if (user && token) {
        deleteGame(gameId, token)
          .then(({ status }) => {
            if (status === 204) {
              const newGames = games.filter(({ id }) => id !== gameId)
              setGames(newGames)
              setFlashProps({
                hidden: false,
                type: 'success',
                message: 'Success! Your game has been deleted.',
              })
            }
          })
          .catch((e: ApiError) => {
            if (import.meta.env.DEV)
              console.log(`Error ${e.name}: ${e.message}`)

            const message =
              e.code === 404 ? NOT_FOUND_MESSAGE : UNEXPECTED_ERROR_MESSAGE

            setFlashProps({
              hidden: false,
              type: 'error',
              message,
            })
          })
      }
    },
    [user, token, games]
  )

  const value = {
    games,
    gamesLoadingState,
    destroyGame,
  }

  useEffect(() => {
    requireLogin()
  }, [requireLogin])

  useEffect(() => {
    if (authLoading) return

    fetchGames()
  }, [authLoading, fetchGames])

  return <GamesContext.Provider value={value}>{children}</GamesContext.Provider>
}
