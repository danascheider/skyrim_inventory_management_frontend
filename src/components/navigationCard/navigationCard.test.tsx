import { describe, test, expect } from 'vitest'
import { renderWithRouter } from '../../setupTests'
import { BLUE } from '../../utils/colorSchemes'
import { ColorProvider } from '../../contexts/colorContext'
import paths from '../../routing/paths'
import NavigationCard from './navigationCard'

describe('NavigationCard component', () => {
  test('NavigationCard component displays properly', () => {
    const wrapper = renderWithRouter(
      <ColorProvider colorScheme={BLUE}>
        <NavigationCard href={paths.home}>Go Home!</NavigationCard>
      </ColorProvider>
    )
    expect(wrapper).toBeTruthy()

    const a = wrapper.container.querySelector('a')
    expect(a?.textContent).toBe('Go Home!')
    expect(a?.href).toBe(paths.home)
  })

  test('matches snapshot', () => {
    const wrapper = renderWithRouter(
      <ColorProvider colorScheme={BLUE}>
        <NavigationCard href={paths.home}>Go Home!</NavigationCard>
      </ColorProvider>
    )

    expect(wrapper).toMatchSnapshot()
  })
})
