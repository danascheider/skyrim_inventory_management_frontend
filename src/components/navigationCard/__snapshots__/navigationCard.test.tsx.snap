// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NavigationCard component > matches snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <a
        class="_root_396ea1"
        href="/"
        style="--background-color: #2274a5; --hover-color: #1e6894; --text-color: #fff;"
      >
        Go Home!
      </a>
    </div>
  </body>,
  "container": <div>
    <a
      class="_root_396ea1"
      href="/"
      style="--background-color: #2274a5; --hover-color: #1e6894; --text-color: #fff;"
    >
      Go Home!
    </a>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
